### PR TITLE
fix: add support for mongodb driver v5

### PR DESCRIPTION
**Pull request by MatheusMeloAntiquera**
This new fix allow support to mongodb driver v5

Closes: #7907

### Description of change

Add support for mongodb v5.0.1

- `mapReduce` was removed because is deprecated since 5.0, instead should use an aggregation pipeline
(https://www.mongodb.com/docs/manual/reference/method/db.collection.mapReduce/)

### Pull-Request Checklist

- [x] Code is up-to-date with the `master` branch
- [x] `npm run format` to apply prettier formatting
- [x] `npm run test` passes with this change
- [x] This pull request links relevant issues as `Fixes #0000`
- [x] There are new or updated unit tests validating the change
- [x] Documentation has been updated to reflect this change
- [x] The new commits follow conventions explained in [CONTRIBUTING.md](https://github.com/typeorm/typeorm/blob/master/CONTRIBUTING.md)

### CODE REVIEW

**Comment by tgrassl** [comment]
Hi, are there any updates on this or is there something i could help out with? Would love to get support for the new driver soon 😃

**Comment by tgrassl** [comment]
Hi, i looked through the code and it looks like this behaviour is related to this issue: https://github.com/typeorm/typeorm/issues/9727

The additional `.save()` calls always results in an insert because the condition of whether it should be updated never matches. This is because the resulting transformed entities of the `findByIds` query in `SubjectDatabaseEntityLoader` don't contain an id property but the subject entity does. https://github.com/typeorm/typeorm/blob/58fc08840a4a64ca1935391f4709a784c3f0b373/src/persistence/SubjectDatabaseEntityLoader.ts#L116
Entity from DB: CleanShot 2023-03-05 at 18 47 18@2x
Subject Entity to persist: CleanShot 2023-03-05 at 19 03 25@2x

Because it has no id the following happens:

1. [findByPersistEntityLike](https://github.com/typeorm/typeorm/blob/58fc08840a4a64ca1935391f4709a784c3f0b373/src/persistence/SubjectDatabaseEntityLoader.ts#L163) fails to match the result to the subject
2. subject.databaseEntity is skipped
3. subject.databaseEntityLoaded is set to true
4. SubjectExecutor only finds a subject to be inserted instead of updated because it has no databaseEntity

https://github.com/typeorm/typeorm/blob/58fc08840a4a64ca1935391f4709a784c3f0b373/src/persistence/SubjectExecutor.ts#L134-L148
https://github.com/typeorm/typeorm/blob/58fc08840a4a64ca1935391f4709a784c3f0b373/src/persistence/Subject.ts#L182-L200

I think the fix would be to check for the `propertyName` of the ObjectIdColumn instead of only the `databaseNameWithoutPrefixes` in the DocumentEntityTransformer [transform](https://github.com/typeorm/typeorm/blob/58fc08840a4a64ca1935391f4709a784c3f0b373/src/query-builder/transformer/DocumentToEntityTransformer.ts#L29):
```typescript
...
        // handle _id property the special way
        if (metadata.objectIdColumn) {
            const { databaseNameWithoutPrefixes, propertyName } =
                metadata.objectIdColumn

            const documentIdWithoutPrefixes =
                document[databaseNameWithoutPrefixes]
            const documentIdWithPropertyName = document[propertyName]

            if (documentIdWithoutPrefixes) {
                entity[propertyName] = documentIdWithoutPrefixes
                hasData = true
            } else if (documentIdWithPropertyName) {
                entity[propertyName] = documentIdWithPropertyName
                hasData = true
            }
        }
...
```

This now results in a update on the same entity instead of an insert. I pushed my changes and created a pr for your branch 👍

**Comment by MatheusMeloAntiquera** [comment]
[tgrassl](https://github.com/tgrassl)

Thank you for your help dude!

**Comment by MatheusMeloAntiquera** [comment]
@pleerock
Can you do the review please? 

There is an error Oracle test, but we didn't any changes on Oracle source

**Comment by tgrassl** [comment]
@pleerock I would like to update the typings, so everyone can now use the official typings already included in the `mongodb` package. This would also finally remove the confusion around [ObjectId and ObjectID](https://github.com/mongodb/node-mongodb-native/blob/HEAD/etc/notes/CHANGES_5.0.0.md#objectid-type-removed-in-favor-of-objectid) (e.g. https://github.com/typeorm/typeorm/issues/2238)

As far i can see, i cannot use `import { ObjectId } from 'mongodb'` inside the following files, because that would require non-mongodb users to install mongodb right?

- EntityManager.ts
- Repository.ts
- BaseEntity.ts
- EntityId.ts
- FindOptionsOrder.ts
- FindOptionsRelations.ts
- FindOptionsSelect.ts
- FindOptionsWhere.ts

These all just need the type of the ObjectId. Mongo specific files like MongoEntityManager are already imported from mongodb, so I guess they are ok.
My idea would be to only copy the [ObjectId typings](https://github.com/mongodb/js-bson/blob/main/src/objectid.ts) into typeorm but not export it from the package, so these files can use that internal type. Would that be fine for you?

File: src/entity-manager/MongoEntityManager.ts [hunk 1]
```diff
@@ -1,62 +1,76 @@
...
+    InsertOneOptions,
+    InsertOneResult,
+    InsertManyResult,
+    UnorderedBulkOperation,
+    OrderedBulkOperation,
+    IndexInformationOptions,
+    ObjectId,
+    MongoError,
+    FilterOperators,
+} from "mongodb"
```
Comment (pleerock) [inline src/entity-manager/MongoEntityManager.ts:64]: we can't have direct imports from mongodb

**Comment by pleerock** [comment]
> These all just need the type of the ObjectId. Mongo specific files like MongoEntityManager already import from mongodb, so I guess they are ok.

No, we must not have imports from `mongodb` anywhere, because non-monogdb users will have tsc errors and they'll be forced to install mongodb package even if they don't use it. That's bad.

**Comment by mptr** [comment]
@MatheusMeloAntiquera how about not submitting the same changes in separate PRs?
I've added you and @tgrassl to my fork. May we close this PR then?
(Also it would be nice to not commit my changes just by copying them on your branch. Instead pull the commits of collaborators on your branch to keep a clean commit history.)

**Comment by MatheusMeloAntiquera** [comment]
> @MatheusMeloAntiquera how about not submitting the same changes in separate PRs? I've added you and @tgrassl to my fork. May we close this PR then? (Also it would be nice to not commit my changes just by copying them on your branch. Instead pull the commits of collaborators on your branch to keep a clean commit history.)

I will keep this PR, sorry for copy your code I wanted to resolve this quickly, but you should have create a PR to my branch, not directly to typeorm repository

**Pull request closed without merging**